### PR TITLE
CLN: remove Index.__inv__ , add Index.__invert__

### LINKED ISSUE #22335: Remove Index.__inv__
#### Code Sample, a copy-pastable example if possible

```python
In [1]: import pandas as pd

In [2]: idx = pd.Index([2, 4, 6])

In [3]: idx.__inv__()
Out[3]: Int64Index([-2, -4, -6], dtype='int64')
```
#### Problem description

``idx.__inv__`` does what ``idx.__neg__`` does, and not do what its name suggests (that is, ``operator.inv``). By the way, ``~idx`` raises ``TypeError`` anyway, and ``idx.__invert__`` does not exist, differently from what [the Python docs suggest](https://docs.python.org/2/library/operator.html).

Notice that instead
```
In [4]: s = pd.Series([2, 4, 6])

In [5]: ~s
Out[5]: 
0   -3
1   -5
2   -7
dtype: int64

In [6]: s.__invert__()
Out[6]: 
0   -3
1   -5
2   -7
dtype: int64
```
which is correct (and ``Series.__inv__`` does not exist).

#### Expected Output

```
Out[3]: Int64Index([-3, -5, -7], dtype='int64')
```

Given how wrong, undocumented and not exposed this is, I suggest to immediately remove ``Index.__inv__``. Then we can later decide to introduce ``Index.__invert__`` which does the right thing, and also enable the ``~`` operator.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-6-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.24.0.dev0+460.geb0ac5437
pytest: 3.5.0
pip: 9.0.1
setuptools: 39.2.0
Cython: 0.28.4
numpy: 1.14.3
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.5.6
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: 1.2.0dev
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.2.2.post1634.dev0+ge8120cf6d
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.3.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #22335
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

@jreback @jorisvandenbossche suggestions on where to put a test (``pandas/tests/test_arithmetic.py`` would seem the right place, except it only contains tests involving datetimelike) and which version to target (not sure 0.23.5 is OK)?

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
> which version to target (not sure 0.23.5 is OK)?

In general, if something is not a (critical) bug fix, it should never be considered for a bug fix release (and certainly not something that is more an API clean-up)

**Comment by jorisvandenbossche** [comment]
> something that is more an API clean-up

Speaking of which, shouldn't we rather deprecate this first before removing?

**Comment by jorisvandenbossche** [comment]
> shouldn't we rather deprecate this first before removing?

OK, didn't yet read your issue, so missed that `~` is currently not working. Then it is fine removing it without deprecation.

**Comment by toobaz** [comment]
@jreback @jorisvandenbossche while I was at it, given that this is going to 0.24 anyway, I thought it wouldn't harm to actually enable ``__invert__``, consistently with the case of ``Series``.

File: pandas/tests/test_arithmetic.py [hunk 1]
```diff
@@ -73,6 +73,41 @@ def box_df_fail(request):
     return request.param
 
 
+# ------------------------------------------------------------------
+# Numeric types Arithmetic Operations
+
+class TestNumericOps(object):
+
+    @pytest.mark.parametrize('opname', ['neg', 'pos', 'abs', 'inv', 'invert'])
```
Comment (jreback) [inline pandas/tests/test_arithmetic.py]: can you add this in conftest as a fixture (unary_op_fixtures)?

File: pandas/tests/test_arithmetic.py [hunk 2]
```diff
@@ -73,6 +73,41 @@ def box_df_fail(request):
...
+        # GH 22335
+        data = [-2, 0, 4, 5]
+        obj = box(data, dtype=dtype)
+
+        def call_method():
+            return getattr(obj, '__{}__'.format(opname))()
+
+        if opname == 'inv':
+            # GH 22335 - 'inv' was wrong and was removed
+            pytest.raises(AttributeError, call_method)
```
Comment (jreback) [inline pandas/tests/test_arithmetic.py]: make these if/elseif, then don't need the return (put the final clause in the else)

**Review by jreback — CHANGES_REQUESTED** [review]
can you also check that the ops are correctly defined in pandas/core/arrays/base.py (for EA arrays).

**Review by jorisvandenbossche — APPROVED** [review]
Apart from Jeff's comments, looks good

**Comment by jreback** [comment]
@toobaz if you can rebase and check extension arrays

**Comment by toobaz** [comment]
> @toobaz if you can rebase and check extension arrays

@jreback I'm afraid I can do that only in September, but while I'm at it:

> can you add this in conftest as a fixture (unary_op_fixtures)?

In the general ``pandas/conftest.py`` ?

> can you also check that the ops are correctly defined in pandas/core/arrays/base.py (for EA arrays).

Do you mean adding EA arrays to ``box`` fixture? Manually or in some automated way?

(In any case, if some tests fail I would defer to another PR)

**Comment by jreback** [comment]
closing as stale. if you want to continue, pls merge master and re-open.

**Pull request closed without merging**